File: javascript/src/lcs.js
const lcs = (s1, s2) => {
  let res = ['', '']
  let i = -1, j = -1
  let set = new Set([...s2]) 

  while(++i < s1.length) {
    if (j+1 >= s2.length) {
      j = -1
      res = [res[0].length > res[1].length ? res[0] : res[1], '']
    }

    if (set.has(s1[i])) {
      while(++j < s2.length) {
        if (s1[i] === s2[j]) {
          res[1] += s1[i]
          break;
        }
      }
    }
  }

  return res.pop()
}
console.log(lcs('AGGTAB', 'GXTXAYB'))
console.log(lcs('ABCDE', 'BCDEA'))
console.log(lcs('ABCDE', 'FGHIJ'))

File: python/src/lcs.py
def lcs(s1,s2):
    res = ['','']
    i, j = 0, 0
    memo = set(s2)

    while i < len(s1):
        if j == len(s2):
            j = 0
            res = [max(res, key=len), '']
        if s1[i] in memo:
            while j < len(s2):
                if s1[i] == s2[j]:
                    res[1] += s1[i]
                    j += 1
                    break
                else: j += 1
        i += 1    
    return res.pop()

print(lcs('AGGTAB', 'GXTXAYB'))
print(lcs('ABCDE', 'BCDEA'))
print(lcs('ABCDE', 'FGHIJ'))
